### PR TITLE
Mafia: add game and host bans

### PR DESCRIPTION
**Pull request by whalemer**

### CODE REVIEW

**Comment by whalemer** [comment]
not sure how this passed locally

File: chat-plugins/mafia.js [hunk 1]
```diff
@@ -47,59 +49,109 @@
...
-		JSON.stringify(logs)
-	));
+// Load bans
+
+bans = readFile(BANS_FILE);
+if (!bans) bans = {gameBans: {}, hostBans: {}};
+const now = Date.now();
+
+for (const userid in bans.gameBans) {
+	if (bans.gameBans[userid] < now) {
```
Comment (ZestOfLife) [inline chat-plugins/mafia.js]: Shouldnt it be <=?

File: chat-plugins/mafia.js [hunk 2]
```diff
@@ -47,59 +49,109 @@
...
+		delete bans.gameBans[userid];
+	}
+}
+for (const userid in bans.hostBans) {
+	if (bans.hostBans[userid] < now) {
```
Comment (ZestOfLife) [inline chat-plugins/mafia.js]: Same here

File: chat-plugins/mafia.js [hunk 3]
```diff
@@ -47,59 +49,109 @@
...
+
+/**
+ *
+ * @param {string} userid
+ * @param {boolean} host
+ */
+function isBanned(userid, host = false) {
+	const banTable = host ? bans.hostBans : bans.gameBans;
+	if (!(userid in banTable)) return false;
+	if (banTable[userid] < Date.now()) {
```
Comment (ZestOfLife) [inline chat-plugins/mafia.js]: and here

File: chat-plugins/mafia.js [hunk 4]
```diff
@@ -1542,19 +1597,20 @@ const commands = {
 			switch (args[0]) {
 			case 'forceadd':
 			case 'add':
+				if (!this.canTalk()) return;
 				let targetUser = Users(args[1]);
 				if ((!targetUser || !targetUser.connected) && args[0] !== 'forceadd') return this.errorReply(`User ${args[1]} not found. To forcefully add the user to the queue, use /mafia queue forceadd, ${args[1]}`);
 				if (hostQueue.includes(args[1])) return this.errorReply(`User ${args[1]} is already on the host queue.`);
 				hostQueue.push(args[1]);
-				this.sendReply(`User ${args[1]} has been added to the host queue.`);
+				room.add(`User ${args[1]} has been added to the host queue.`).update();
```
Comment (ZestOfLife) [inline chat-plugins/mafia.js]: If we are broadcasting who is getting added in the host queue, make sure to announce who added them in the queue too I think so.

File: chat-plugins/mafia.js [hunk 5]
```diff
@@ -1542,19 +1597,20 @@ const commands = {
...
 				break;
 			case 'del':
 			case 'delete':
 			case 'remove':
 				let index = hostQueue.indexOf(args[1]);
 				if (index === -1) return this.errorReply(`User ${args[1]} is not on the host queue.`);
 				hostQueue.splice(index, 1);
-				this.sendReply(`User ${args[1]} has been removed from the host queue`);
+				room.add(`User ${args[1]} has been removed from the host queue`).update();
```
Comment (ZestOfLife) [inline chat-plugins/mafia.js]: Probably the same here

File: chat-plugins/mafia.js (hunk 1 above)
Comment (whalemer) [reply chat-plugins/mafia.js]: i don't think that literally a millisecond will make a difference

File: chat-plugins/mafia.js [hunk 6]
```diff
@@ -2403,6 +2451,38 @@ const commands = {
...
 		],
 
+		unhostban: 'hostban',
+		hostban: function (target, room, user, connection, cmd) {
+			if (!room || !room.mafiaEnabled) return this.errorReply(`Mafia is disabled for this room.`);
+			if (room.id !== 'mafia') return this.errorReply(`This command can only be used in the Mafia room.`);
+
+			const duration = parseInt(this.splitTarget(target, false));
+			if (!this.can('ban', this.targetUser, room)) return false;
+			if (!this.targetUser) return this.errorReply(`User ${target} not found.`);
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: Check this before checking permissions

File: chat-plugins/mafia.js [hunk 7]
```diff
@@ -2403,6 +2451,38 @@ const commands = {
...
+
+			const isUnban = (cmd.startsWith('un'));
+			if (isHostBanned(toId(this.targetUsername)) === !isUnban) return this.errorReply(`${this.targetUsername} is ${isUnban ? 'not' : 'already'} banned from hosting games.`);
+
+			if (isUnban) {
+				delete hostBans[toId(this.targetUsername)];
+				this.modlog(`MAFIAUNHOSTBAN`, this.targetUser);
+			} else {
+				if (isNaN(duration) || duration < 1) return this.parse('/help mafia gameban');
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: hostban not gameban

File: chat-plugins/mafia.js [hunk 8]
```diff
@@ -2403,6 +2451,38 @@ const commands = {
 			`/mafia [hostlost|playlogs] - View the host logs or play logs for the current or last month. Requires % @ * # & ~`,
 		],
 
+		unhostban: 'hostban',
+		hostban: function (target, room, user, connection, cmd) {
+			if (!room || !room.mafiaEnabled) return this.errorReply(`Mafia is disabled for this room.`);
+			if (room.id !== 'mafia') return this.errorReply(`This command can only be used in the Mafia room.`);
+
+			const duration = parseInt(this.splitTarget(target, false));
+			if (!this.can('ban', this.targetUser, room)) return false;
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: Let drivers hostban too.

File: chat-plugins/mafia.js [hunk 9]
```diff
@@ -2403,6 +2451,38 @@ const commands = {
...
+				if (isNaN(duration) || duration < 1) return this.parse('/help mafia gameban');
+				if (duration > 7) return this.errorReply(`Bans cannot be longer than 7 days.`);
+
+				hostBans[toId(this.targetUsername)] = Date.now() + 1000 * 60 * 60 * 24 * duration;
+				this.modlog(`MAFIAHOSTBAN`, this.targetUser, `for ${duration} days.`);
+				const queueIndex = hostQueue.indexOf(toId(this.targetUsername));
+				if (queueIndex > -1) hostQueue.splice(queueIndex, 1);
+			}
+			writeFile(BANS_FILE, hostBans);
+			room.add(`${this.targetUsername} was ${isUnban ? 'un' : ''}banned from hosting games by ${user}${!isUnban ? ` for ${duration} days` : ''}.`).update();
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: `by ${user.name}`

File: chat-plugins/mafia.js [hunk 10]
```diff
@@ -2403,6 +2451,38 @@ const commands = {
...
+		},
+		gamebanhelp: [
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: hostban

also add %'s to the requires list.

File: chat-plugins/mafia.js [hunk 11]
```diff
@@ -47,59 +48,97 @@
...
+ * @param {string} name
+ */
+function readFile(name) {
+	try {
+		const json = FS(name).readIfExistsSync();
+		if (!json) {
+			writeFile(name, "{}");
+			return false;
+		}
+		return Object.assign(Object.create(null), JSON.parse(json));
```
Comment (Zarel) [inline chat-plugins/mafia.js]: If we bump the minimum Node version, we could start using `{_proto_: null, ...JSON.parse(json)}`

Comment (HoeenCoder) [reply chat-plugins/mafia.js]: How far would we have to bump it? And would it be worth the hassle of raising PS's min node version?

**Comment by whalemer** [comment]
also, is ``{__proto__: null, whatever}`` preferred over ``Object.assign(Object.create(null), {whatever})``

File: chat-plugins/mafia.js (hunk 11 above)
Comment (Zarel) [reply chat-plugins/mafia.js]: I think 8.2 is where object-rest-spread was introduced. I was mostly shitposting there, though. I don't have a strong opinion on which syntax is better, although I do in general prefer doing things syntactically to calling functions; I think it makes things more readable.

**Pull request merged**